File: claudia-deployment/mcp-client-example.py
#!/usr/bin/env python3
"""
Claude Code MCP Client 示例
演示如何连接到 Claude Code MCP Server 并使用其功能
"""

import subprocess
import json
import sys
import time

class ClaudeCodeMCPClient:
    def __init__(self, claude_path="claude"):
        """初始化 MCP 客户端"""
        self.claude_path = claude_path
        self.process = None
        self.request_id = 0
    
    def start_server(self):
        """启动 Claude Code MCP Server"""
        print("🚀 启动 Claude Code MCP Server...")
        try:
            self.process = subprocess.Popen(
                [self.claude_path, "mcp", "serve"],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                bufsize=1  # 行缓冲
            )
            # 等待服务器启动
            time.sleep(2)
            print("✅ MCP Server 启动成功")
            return True
        except Exception as e:
            print(f"❌ 启动失败: {e}")
            return False
    
    def send_request(self, method, params=None):
        """发送 MCP 请求"""
        if not self.process:
            raise Exception("MCP Server 未启动")
        
        self.request_id += 1
        request = {
            "jsonrpc": "2.0",
            "id": self.request_id,
            "method": method,
            "params": params or {}
        }
        
        print(f"📤 发送请求: {method}")
        try:
            # 发送请求
            request_json = json.dumps(request) + "\n"
            self.process.stdin.write(request_json)
            self.process.stdin.flush()
            
            # 读取响应
            response_line = self.process.stdout.readline()
            if response_line:
                response = json.loads(response_line.strip())
                return response
            else:
                raise Exception("没有收到响应")
        except Exception as e:
            print(f"❌ 请求失败: {e}")
            return None
    
    def list_tools(self):
        """获取可用工具列表"""
        print("\n📋 获取可用工具...")
        response = self.send_request("tools/list")
        if response and "result" in response:
            tools = response["result"]["tools"]
            print(f"✅ 找到 {len(tools)} 个工具:")
            for tool in tools:
                print(f"   - {tool['name']}: {tool.get('description', '无描述')}")
            return tools
        else:
            print("❌ 获取工具列表失败")
            return []
    
    def list_directory(self, path="."):
        """列出目录内容"""
        print(f"\n📁 列出目录: {path}")
        response = self.send_request("tools/call", {
            "name": "LS",
            "arguments": {"path": path}
        })
        if response and "result" in response:
            print("✅ 目录内容:")
            content = response["result"]["content"]
            for item in content:
                print(f"   {item}")
            return content
        else:
            print("❌ 列出目录失败")
            return []
    
    def read_file(self, file_path):
        """读取文件"""
        print(f"\n📄 读取文件: {file_path}")
        response = self.send_request("tools/call", {
            "name": "Read",
            "arguments": {"file_path": file_path}
        })
        if response and "result" in response:
            content = response["result"]["content"][0]["text"]
            print(f"✅ 文件内容 ({len(content)} 字符):")
            # 显示前 200 字符
            preview = content[:200] + "..." if len(content) > 200 else content
            print(f"   {preview}")
            return content
        else:
            print("❌ 读取文件失败")
            return None
    
    def execute_command(self, command):
        """执行命令"""
        print(f"\n💻 执行命令: {command}")
        response = self.send_request("tools/call", {
            "name": "Bash",
            "arguments": {"command": command}
        })
        if response and "result" in response:
            output = response["result"]["content"][0]["text"]
            print(f"✅ 命令输出:")
            print(f"   {output}")
            return output
        else:
            print("❌ 执行命令失败")
            return None
    
    def stop_server(self):
        """停止 MCP Server"""
        if self.process:
            print("\n🛑 停止 MCP Server...")
            self.process.terminate()
            self.process.wait()
            print("✅ MCP Server 已停止")

def main():
    """主函数 - 演示 MCP 客户端使用"""
    print("🎯 Claude Code MCP Client 示例")
    print("=" * 50)
    
    # 检查命令行参数
    claude_path = sys.argv[1] if len(sys.argv) > 1 else "claude"
    
    # 创建客户端
    client = ClaudeCodeMCPClient(claude_path)
    
    try:
        # 启动服务器
        if not client.start_server():
            sys.exit(1)
        
        # 获取工具列表
        tools = client.list_tools()
        
        # 如果有 LS 工具，列出当前目录
        if any(tool["name"] == "LS" for tool in tools):
            client.list_directory(".")
        
        # 如果有 Read 工具，读取一个文件
        if any(tool["name"] == "Read" for tool in tools):
            # 尝试读取 README.md 或 package.json
            for filename in ["README.md", "package.json", "requirements.txt"]:
                client.read_file(filename)
                break
        
        # 如果有 Bash 工具，执行一个简单命令
        if any(tool["name"] == "Bash" for tool in tools):
            client.execute_command("pwd")
            client.execute_command("date")
        
        print("\n🎉 示例演示完成！")
        
    except KeyboardInterrupt:
        print("\n⏹️  用户中断")
    except Exception as e:
        print(f"\n❌ 发生错误: {e}")
    finally:
        client.stop_server()

if __name__ == "__main__":
    main()

File: claudia-deployment/mcp-client-example.js
#!/usr/bin/env node
/**
 * Claude Code MCP Client 示例 (Node.js)
 * 演示如何连接到 Claude Code MCP Server 并使用其功能
 */

const { spawn } = require('child_process');
const readline = require('readline');

class ClaudeCodeMCPClient {
    constructor(claudePath = 'claude') {
        this.claudePath = claudePath;
        this.process = null;
        this.requestId = 0;
        this.pendingRequests = new Map();
    }

    async startServer() {
        console.log('🚀 启动 Claude Code MCP Server...');
        try {
            this.process = spawn(this.claudePath, ['mcp', 'serve']);
            
            // 设置输出处理
            this.rl = readline.createInterface({
                input: this.process.stdout,
                crlfDelay: Infinity
            });

            // 处理响应
            this.rl.on('line', (line) => {
                try {
                    const response = JSON.parse(line);
                    if (response.id && this.pendingRequests.has(response.id)) {
                        const resolve = this.pendingRequests.get(response.id);
                        this.pendingRequests.delete(response.id);
                        resolve(response);
                    }
                } catch (e) {
                    console.error('解析响应失败:', e);
                }
            });

            // 错误处理
            this.process.stderr.on('data', (data) => {
                console.error('MCP Server 错误:', data.toString());
            });

            // 等待服务器启动
            await new Promise(resolve => setTimeout(resolve, 2000));
            console.log('✅ MCP Server 启动成功');
            return true;
        } catch (e) {
            console.error('❌ 启动失败:', e);
            return false;
        }
    }

    async sendRequest(method, params = {}) {
        if (!this.process) {
            throw new Error('MCP Server 未启动');
        }

        this.requestId++;
        const request = {
            jsonrpc: "2.0",
            id: this.requestId,
            method: method,
            params: params
        };

        console.log(`📤 发送请求: ${method}`);
        
        return new Promise((resolve, reject) => {
            // 存储 promise resolver
            this.pendingRequests.set(this.requestId, resolve);
            
            // 发送请求
            const requestJson = JSON.stringify(request) + '\n';
            this.process.stdin.write(requestJson);
            
            // 设置超时
            setTimeout(() => {
                if (this.pendingRequests.has(this.requestId)) {
                    this.pendingRequests.delete(this.requestId);
                    reject(new Error('请求超时'));
                }
            }, 10000);
        });
    }

    async listTools() {
        console.log('\n📋 获取可用工具...');
        try {
            const response = await this.sendRequest('tools/list');
            if (response.result) {
                const tools = response.result.tools;
                console.log(`✅ 找到 ${tools.length} 个工具:`);
                tools.forEach(tool => {
                    console.log(`   - ${tool.name}: ${tool.description || '无描述'}`);
                });
                return tools;
            }
        } catch (e) {
            console.error('❌ 获取工具列表失败:', e);
        }
        return [];
    }

    async listDirectory(path = '.') {
        console.log(`\n📁 列出目录: ${path}`);
        try {
            const response = await this.sendRequest('tools/call', {
                name: 'LS',
                arguments: { path: path }
            });
            if (response.result) {
                console.log('✅ 目录内容:');
                const content = response.result.content;
                content.forEach(item => {
                    console.log(`   ${item.text || item}`);
                });
                return content;
            }
        } catch (e) {
            console.error('❌ 列出目录失败:', e);
        }
        return [];
    }

    async readFile(filePath) {
        console.log(`\n📄 读取文件: ${filePath}`);
        try {
            const response = await this.sendRequest('tools/call', {
                name: 'Read',
                arguments: { file_path: filePath }
            });
            if (response.result && response.result.content) {
                const content = response.result.content[0].text;
                console.log(`✅ 文件内容 (${content.length} 字符):`);
                // 显示前 200 字符
                const preview = content.length > 200 ? 
                    content.substring(0, 200) + '...' : content;
                console.log(`   ${preview}`);
                return content;
            }
        } catch (e) {
            console.error('❌ 读取文件失败:', e);
        }
        return null;
    }

    async executeCommand(command) {
        console.log(`\n💻 执行命令: ${command}`);
        try {
            const response = await this.sendRequest('tools/call', {
                name: 'Bash',
                arguments: { command: command }
            });
            if (response.result && response.result.content) {
                const output = response.result.content[0].text;
                console.log('✅ 命令输出:');
                console.log(`   ${output}`);
                return output;
            }
        } catch (e) {
            console.error('❌ 执行命令失败:', e);
        }
        return null;
    }

    async searchFiles(pattern, path = '.') {
        console.log(`\n🔍 搜索文件: ${pattern} in ${path}`);
        try {
            const response = await this.sendRequest('tools/call', {
                name: 'Glob',
                arguments: { pattern: pattern, path: path }
            });
            if (response.result && response.result.content) {
                const files = response.result.content;
                console.log(`✅ 找到 ${files.length} 个文件:`);
                files.forEach(file => {
                    console.log(`   ${file.text || file}`);
                });
                return files;
            }
        } catch (e) {
            console.error('❌ 搜索文件失败:', e);
        }
        return [];
    }

    stopServer() {
        if (this.process) {
            console.log('\n🛑 停止 MCP Server...');
            this.process.kill();
            console.log('✅ MCP Server 已停止');
        }
    }
}

async function main() {
    console.log('🎯 Claude Code MCP Client 示例 (Node.js)');
    console.log('='.repeat(50));

    // 检查命令行参数
    const claudePath = process.argv[2] || 'claude';
    
    // 创建客户端
    const client = new ClaudeCodeMCPClient(claudePath);

    try {
        // 启动服务器
        if (!await client.startServer()) {
            process.exit(1);
        }

        // 获取工具列表
        const tools = await client.listTools();
        const toolNames = tools.map(t => t.name);

        // 演示各种功能
        if (toolNames.includes('LS')) {
            await client.listDirectory('.');
        }

        if (toolNames.includes('Glob')) {
            await client.searchFiles('*.json');
        }

        if (toolNames.includes('Read')) {
            // 尝试读取一些常见文件
            const filesToTry = ['package.json', 'README.md', 'requirements.txt', '.gitignore'];
            for (const file of filesToTry) {
                const content = await client.readFile(file);
                if (content) break; // 找到一个就够了
            }
        }

        if (toolNames.includes('Bash')) {
            await client.executeCommand('pwd');
            await client.executeCommand('ls -la | head -5');
            await client.executeCommand('date');
        }

        console.log('\n🎉 示例演示完成！');

    } catch (e) {
        if (e.message === 'SIGINT') {
            console.log('\n⏹️  用户中断');
        } else {
            console.error('\n❌ 发生错误:', e);
        }
    } finally {
        client.stopServer();
    }
}

// 处理 Ctrl+C
process.on('SIGINT', () => {
    console.log('\n⏹️  接收到中断信号...');
    process.exit(0);
});

// 运行主函数
if (require.main === module) {
    main().catch(console.error);
}